### DISCUSSION TITLE
Risk of hitting CMS rate limit during build step?

### DISCUSSION THREAD

**Discussion by neallseth** [vercel/vercel#5185] [Help]
I have a Next.js app deployed to Vercel, which is statically generated at build-time.  The build depends on a CMS with a rate limit of 55 requests per second (Contentful), and it's entirely possible that this will be exceeded during the SSG build process.
What patterns can be used to ensure a successful build, given the limit?

I'm sure this is a common problem for large, CMS-dependent sites.  Does Vercel have any way of detecting and retrying failed requests?

Would really appreciate any advice or info on this!

Thanks!

**Comment by ChrisDalley** [comment]
I'm encountering some similar issues with another 3rd Party API and a static site generation with NextJS. Locally I have added timeouts and api request retry logic on 429 responses and this seems to work but when pushing up to Vercel the pages all build in parallel rather than sequentially so I every 2-3 builds seem to hit the rate limit and the build fails. I'm still looking for a solution to this but there must be one considering the number of large headless CMS static sites hosted on Vercel...
